pkg/allocateip/allocateip.go

```diff
@@ -93,25 +99,54 @@ func ensureHostTunnelAddress(ctx context.Context, c client.Interface, nodename s
 		logCtx.WithError(err).Fatalf("Unable to retrieve tunnel address. Error getting node '%s'", nodename)
 	}
 
-	// Get the address
+	// Get the address and ipam attribute string
 	var addr string
+	var attrString string
 	if vxlan {
 		addr = node.Spec.IPv4VXLANTunnelAddr
+		attrString = vxlanIPAMAttrString
 	} else if node.Spec.BGP != nil {
 		addr = node.Spec.BGP.IPv4IPIPTunnelAddr
+		attrString = ipipIPAMAttrString
 	}
 
+	// Work out if we need to assign a tunnel address.
+	var assign, release bool
 	if addr == "" {
 		// The tunnel has no IP address assigned, assign one.
-		logCtx.Debug("tunnel is not assigned - assign IP")
-		assignHostTunnelAddr(ctx, c, nodename, cidrs, vxlan)
-	} else if isIpInPool(addr, cidrs) {
-		// The tunnel address is still valid, so leave as it.
-		logCtx.WithField("IP", addr).Info("tunnel address is still valid")
-	} else {
+		logCtx.WithField("Node", nodename).Info("Assign a new tunnel address")
+		release = false
+		assign = true
+	} else if !isIpInPool(addr, cidrs) {
 		// The address that is currently assigned is no longer part
 		// of an encapsulatin-enabled pool, so release the IP, and reassign.
-		logCtx.WithField("IP", addr).Info("Reassigning tunnel address")
+		logCtx.WithField("currentAddr", addr).Info("Current address is not in ippool, reassign")
+		release = true
+		assign = true
+	} else {
+		// The tunnel address is still part of an an encapsulatin-enabled pool.
+		ipAddr := gnet.ParseIP(addr)
+		if err != nil {
+			logCtx.WithError(err).Fatalf("Failed to parse the CIDR '%s'", addr)
+		}
+
+		// Check if we got correct assignment attributes
+		attr, attrErr := c.IPAM().GetAssignmentAttributes(ctx, net.IP{IP: ipAddr})
+		if attrErr == nil && attr[ipam.AttributeType] == attrString {
+			// The tunnel address is still valid, so leave as it.
+			logCtx.WithField("currentAddr", addr).Info("Current address is valid, do nothing")
+			release = false
+			assign = false
+		} else {
```

Comment:
So, I think this covers all the cases we care about. Spelling them out for my own sanity:

- Address is assigned, and has the right attribute - we don't need to do anything
- Address is assigned, but has been allocated to someone else - we need to assign a new address. 
- Address is not assigned, we should allocate a new one and update the node object.

Reply:
Yes. These are cases we try to cover.

Reply:
Also there are two types of `Address is not assigned`, in both cases, we need to allocate a new address and update node object.
- Tunnel address on node object is empty.
- Tunnel address has been set on node object, but it is a free IP address in datastore.